File: app/src/main/java/com/kstechsolutions/mvvm/data/di/modules/ViewModelModule.java
package com.kstechsolutions.mvvm.data.di.modules;

import android.arch.lifecycle.ViewModelProvider;

import com.kstechsolutions.mvvm.modules.base.ViewModelProviderFactory;
import com.kstechsolutions.mvvm.modules.comments.viewmodel.CommentViewModel;
import com.kstechsolutions.mvvm.modules.posts.viewmodel.PostListViewModel;

import dagger.Module;
import dagger.Provides;

@Module
public class ViewModelModule {
    @Provides
    PostListViewModel providePostListViewModel() {
        return new PostListViewModel();
    }

    @Provides
    CommentViewModel provideCommentViewModel() {
        return new CommentViewModel();
    }

    @Provides
    ViewModelProvider.Factory provideBlogViewModel(PostListViewModel postListViewModel) {
        return new ViewModelProviderFactory<>(postListViewModel);
    }

}

File: app/src/main/java/com/kstechsolutions/mvvm/data/di/ViewModelModule.kt
package com.kstechsolutions.mvvm.data.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kstechsolutions.mvvm.data.di.base.ViewModelFactory
import com.kstechsolutions.mvvm.data.di.base.ViewModelKey
import com.kstechsolutions.mvvm.modules.comments.CommentListViewModel
import com.kstechsolutions.mvvm.modules.posts.PostListViewModel
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
abstract class ViewModelModule {
    @Binds
    abstract fun bindViewModelFactory(viewModelFactory: ViewModelFactory): ViewModelProvider.Factory

    @Binds
    @IntoMap
    @ViewModelKey(PostListViewModel::class)
    internal abstract fun postListViewModel(viewModel: PostListViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(CommentListViewModel::class)
    internal abstract fun commentListViewModel(viewModel: CommentListViewModel): ViewModel
}